### PR TITLE
API, Core: Add strict metadata cleanup to TableOperations

### PR DESCRIPTION
**Pull request by amogh-jahagirdar**
Co-authored-by: Ryan Blue <blue@apache.org>

This is a continuation of https://github.com/apache/iceberg/pull/6399

Strict metadata cleanup will only trigger cleanups during commits when the commit fails with an exception which implements the marker interface CleanableFailure This will get used in SnapshotProducer and BaseTransaction and is useful for catalogs like the REST catalog where arbitrary HTTP client exceptions can be thrown (instead of the usual CommitStateUnknown exceptions).

### CODE REVIEW

**Comment by amogh-jahagirdar** [comment]
I still need to write tests for this, putting it up for early eyes on the code

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java [hunk 1]
```diff
@@ -474,6 +461,16 @@ private void commitSimpleTransaction() {
...
+    computeUncommittedFiles();
+
+    // delete all the uncommitted files
+    Tasks.foreach(deletedFiles)
+        .suppressFailureWhenFinished()
+        .onFailure((file, exc) -> LOG.warn("Failed to delete uncommitted file: {}", file, exc))
+        .run(ops.io()::deleteFile);
+  }
+
+  private void computeUncommittedFiles() {
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/BaseTransaction.java:473]: The diff looks a bit strange, but  it's a bit of refactoring which separates

1) computing the files to clean up via going through all the updates and calling `cleanAll` which enqueues the uncommitted files

2.) Actually deleting the files

I did this because in the code there a rereferences to "cleaning up" but really what's happening is an enqueuing of possible files to cleanup after the commit succeeds or fails.

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 2]
```diff
@@ -436,6 +439,12 @@ public void commit() {
     }
   }
 
+  private void cleanupIfPossible(RuntimeException exception) {
+    if (!strictCleanup || exception instanceof CleanableFailure) {
+      Exceptions.suppressAndThrow(exception, this::cleanAll);
+    }
+  }
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:446]: I did a separate helper to avoid checkstyle complaining about cyclomatic complexity.

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 3]
```diff
@@ -399,7 +402,7 @@ public void commit() {
     } catch (CommitStateUnknownException commitStateUnknownException) {
       throw commitStateUnknownException;
     } catch (RuntimeException e) {
-      Exceptions.suppressAndThrow(e, this::cleanAll);
+      cleanupIfPossible(e);
     }
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:406]: This doesn't seem right, we still want the suppress and throw behavior but the cleanup should be conditional on the strictCleanup and CleanableFailure

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java [hunk 4]
```diff
@@ -1006,17 +1014,6 @@ public void testTableAuth(
               .build();
         };
 
-    Mockito.doAnswer(addTableConfig)
```
Comment (amogh-jahagirdar) [inline core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:1009]: shouldn't have been removed

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 5]
```diff
@@ -436,6 +439,14 @@ public void commit() {
...
   }
 
+  private void cleanupAfterFailure(RuntimeException exception) {
+    if (!strictCleanup || exception instanceof CleanableFailure) {
+      Exceptions.suppressAndThrow(exception, this::cleanAll);
+    }
+
+    throw exception;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:447]: This shouldn't throw an exception. The exception should be thrown from the calling code so that it is clear when reading that section how execution will happen. It's fine to pass it in here to determine whether to clean up though.

I'm not sure that this refactoring is worth the method, but it's up to you once this `throw` is moved back.

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java [hunk 6]
```diff
@@ -319,15 +320,7 @@ private void commitCreateTransaction() {
...
-      Tasks.foreach(updates)
-          .suppressFailureWhenFinished()
-          .run(
-              update -> {
-                if (update instanceof SnapshotProducer) {
-                  ((SnapshotProducer) update).cleanAll();
-                }
-              });
-
+      computeUncommittedFiles();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTransaction.java:323]: I think this needs to be refactored more. If the `RuntimeException` was something like a 500 error for a create table or replace table transaction, then the table may have been created or replaced. For strict cleanup, I think we still need to check for a cleanable exception in those cases.

I think these should be:
```java
} catch (RuntimeException e) {
  if (!ops.requiresStrictCleanup() || e instanceof CleanableFailure) {
    cleanUpOnCommitFailure();
  }
}

Tasks.foreach(deletedFiles).run(ops.io()::delete);
```

**Review by rdblue — CHANGES_REQUESTED** [review]
The main issue is that I think the create table and replace table cases also need to respect strict cleanup. Also, there's a refactor that makes the code misleading by throwing an exception unexpectedly from a util method.

@amogh-jahagirdar, in addition to fixing this, I think we also need to look into file cleanup for transactions. I was reviewing the current handling and I noticed that the `deletedFiles` set is never cleared between retries for a `SIMPLE` transaction. As a result, there is a chance that a file that was deleted by one operation's commit but reused when that operation is recommitted will remain in the delete set and be removed.

I think all we need to do is to clear the delete file set before each commit attempt. Could you look into this in a follow-up PR?

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java (hunk 5 above)
Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/SnapshotProducer.java:447]: Good point, it's not a good idea to have a helper method throw a runtime exception, it's less explicit to a reader since they'd have to actually step into the helper. I'll fix  this up

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java (hunk 6 above)
Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/BaseTransaction.java:323]: Right, it's possible that during the create/replace we get a 500 but the create/replace actually was successful. So we still need the cleanable failure check. I realized I'm also missing tests for create/replace/simple transactions so I'll add those

Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/BaseTransaction.java]: I've updated. If I'm not missing anything we would still want to preserve the behavior of cleaning up the files in the `finally` block because that set of deleted files would the files from previous commit attempts (this was already happening regardless of commit outcome, including CommitUnknownException). We should be able to clean those files up in any situation.

**Comment by amogh-jahagirdar** [comment]
Thanks for the review @rdblue, I've updated the PR for updating the create/replace cases to perform the strict metadata cleanup check. One case I wanted to double check was the deletion of files happening in the finally block for create/replace https://github.com/apache/iceberg/pull/8397#discussion_r1316210793 . If I understand right, we still would want to preserve this cleanup regardless of commit outcome or strict cleanup because that set of files would be files from previous commit attempts which we can safely cleanup.


> @amogh-jahagirdar, in addition to fixing this, I think we also need to look into file cleanup for transactions. I was reviewing the current handling and I noticed that the deletedFiles set is never cleared between retries for a SIMPLE transaction. As a result, there is a chance that a file that was deleted by one operation's commit but reused when that operation is recommitted will remain in the delete set and be removed.

I think all we need to do is to clear the delete file set before each commit attempt. Could you look into this in a follow-up PR?

Good catch, sure I will look into this!

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 7]
```diff
@@ -353,6 +356,7 @@ protected TableMetadata refresh() {
   }
 
   @Override
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:359]: It didn't seem worth it to have the exception helper function just to avoid cyclomatic complexity. There's probably better refactoring we could do if we wanted to remove this, for now just decided to suppress warnings.

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java [hunk 8]
```diff
@@ -316,20 +317,13 @@ private void commitCreateTransaction() {
 
     } catch (CommitStateUnknownException e) {
       throw e;
-
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTransaction.java:319]: Nit: unnecessary whitespace change. I also like the whitespace here because it maintains separation.

Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/BaseTransaction.java:319]: Sure, I added back the newline

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java [hunk 9]
```diff
@@ -380,14 +375,9 @@ private void commitReplaceTransaction(boolean orCreate) {
...
-      Tasks.foreach(updates)
-          .suppressFailureWhenFinished()
-          .run(
-              update -> {
-                if (update instanceof SnapshotProducer) {
-                  ((SnapshotProducer) update).cleanAll();
-                }
-              });
+      if (!ops.requireStrictCleanup() || e instanceof CleanableFailure) {
+        cleanUpOnCommitFailure();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTransaction.java:379]: Shouldn't this be `cleanAllUpdates`? Both `cleanUpOnCommitFailure` and the `finally` block are going to run the delete function on every file.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks good to me. Let's merge when tests are passing. Thanks, @amogh-jahagirdar!

File: core/src/main/java/org/apache/iceberg/BaseTransaction.java (hunk 9 above)
Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/BaseTransaction.java]: Oh yes good catch, we're duplicating the call to delete the files. Although I wonder how the create/replace tests for this case were passing.

Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/BaseTransaction.java]: Ah the delete is best effort so the second time when the file no longer exists the delete would fail but we catch it and emit a warn log.

**Comment by amogh-jahagirdar** [comment]
Thanks for review @rdblue ! merging
